libcfnet/client_code.c

```diff
@@ -901,6 +903,18 @@ bool CopyRegularFileNet(const char *source, const char *dest, off_t size,
         }
 
         n_wrote_total += n_read;
+
+        /* A sanity check to make sure we don't write more than the original
+         * filesize acquired by SYNCH ... STAT source. This might happen if
+         * the file was changed while copying. */
+        if (n_wrote_total > size)
+        {
+            Log(LOG_LEVEL_INFO, "Source '%s:%s' changed while copying",
+                conn->this_server, source);
+            close(dd);
+            free(buf);
```

Comment:
It seems to me that you need some more cleanup. Before this and the other failure catch above this one we have

```
        bool w_ok = FileSparseWrite(dd, buf, n_read,
                                    &last_write_made_hole);
```

and so it would seem to me you need more than close/free but rather the same as the other failure case catch above, though I am not sure about conn->error = true but seems also reasonable.

```
            free(buf);
            unlink(dest);
            close(dd);
            FlushFileStream(conn->conn_info->sd, size - n_wrote_total - n_read);
            conn->error = true;
```

Reply:
Good catch 🚀

Reply:
I'm not sure about the `conn->error = true` either. As far as I can see, it's not really checked after executing this code.

Reply:
It should be quite easy to merge the two cleanup blocks into a single one acting on two possible error/failure cases.